chain-vote/src/commitment.rs

```diff
@@ -1,85 +1,61 @@
 use crate::gang::{GroupElement, Scalar};
-use std::ops::{Add, Mul};
-
-/// Pedersen commitment
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct Commitment {
-    c: GroupElement,
-}
+use crate::CRS;
 
+/// Pedersen Commitment key
 #[derive(Clone)]
 pub struct CommitmentKey {
     pub h: GroupElement,
 }
 
 impl CommitmentKey {
+    pub fn to_bytes(&self) -> [u8; GroupElement::BYTES_LEN] {
+        self.h.to_bytes()
+    }
+
+    /// Generate a commitment key from a seed. This function hashes the
+    /// input `buffer`, and creates a group element out of the hash.
     pub fn generate_from_seed(buffer: &mut [u8]) -> Self {
         CommitmentKey {
             h: GroupElement::from_hash(buffer),
         }
     }
-}
-
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
-pub enum Validity {
-    Valid,
-    Invalid,
-}
-
-#[derive(Clone)]
-pub struct Open {
-    m: Scalar,
-    r: Scalar,
-}
 
-impl Commitment {
-    pub const BYTES_LEN: usize = GroupElement::BYTES_LEN;
-
-    pub fn new_open(ck: &CommitmentKey, o: &Open) -> Self {
-        let c = GroupElement::generator() * &o.m + &ck.h * &o.r;
-        Commitment { c }
+    /// Return a commitment with the given opening, `o`
+    pub fn commit_with_open(&self, o: &Open) -> GroupElement {
+        self.commit(&o.m, &o.r)
     }
 
-    pub fn new(ck: &CommitmentKey, m: &Scalar, r: &Scalar) -> Self {
-        let c = GroupElement::generator() * m + &ck.h * r;
-        Commitment { c }
+    /// Return a commitment with the given message, `m`,  and opening key, `r`
+    pub fn commit(&self, m: &Scalar, r: &Scalar) -> GroupElement {
```

Comment:
I thought about this [rng as input]. The problem of a commitment (contrarily to encryption) is that the randomness needs to be stored to memory, as it is needed to prove knowledge of the committed value. So we would need to return the group element and the opening (and still keep this function for verification procedure).

Reply:
And for constructing `Announcement` and `BlindingRandomness` together, it might make sense. I'll give it a try

Reply:
Yep, I was thinking of something like `encrypt_return_r` we use in encryption. Then the `commit` function that accepts the raw randomness instead of a PRNG could be made private and only used internally (for both commit and verificatiol)

Reply:
ah, that is also a good point 👍  I'll do that. Check out the code with the `Announcement` and `BlindingRandomness` as discussed above. I think it is better as you said 👌  (the last commit)